lib/developer_portal/app/controllers/developer_portal/admin/account/invoices_controller.rb

```diff
@@ -1,34 +1,88 @@
+# frozen_string_literal: true
+
 class DeveloperPortal::Admin::Account::InvoicesController < ::DeveloperPortal::BaseController
+  class PaymentGatewayNotSupported < StandardError; end
+
+  rescue_from PaymentGatewayNotSupported, with: :handle_not_found
 
   helper Finance::InvoicesHelper
   helper Accounts::InvoicesHelper
 
   before_action :authorize_finance
   before_action :find_provider
+  before_action :authorize_payment_gateway, only: %i[payment payment_succeeded]
+  before_action :find_invoice, only: %i[show payment payment_succeeded]
+
   activate_menu :account, :invoices
 
   liquify prefix: 'invoices'
 
   def index
-    collection = current_account.invoices.visible_for_buyer.page(params[:page])
-
+    collection = accessible_invoices.page(params[:page])
     invoices = Liquid::Drops::Invoice.wrap(collection)
     pagination = Liquid::Drops::Pagination.new(collection, self)
     assign_drops invoices: invoices, pagination: pagination
   end
 
   def show
-    invoice = current_account.invoices.visible_for_buyer.find(params[:id])
     assign_drops invoice: Liquid::Drops::Invoice.wrap(invoice)
   end
 
+  def payment
+    @payment_intent = @invoice.payment_intents.latest_pending.first!
+    @client_secret = stripe_payment_intent.client_secret
+    @stripe_publishable_key = payment_gateway_options[:publishable_key]
+  end
+
+  def payment_succeeded
+    @payment_intent = @invoice.payment_intents.find_by!(reference: stripe_payment_intent_params[:id])
+
+    service = Finance::StripePaymentIntentUpdateService.new(site_account, stripe_payment_intent)
+    if service.call
+      flash[:notice] = 'Payment transaction updated'
+    else
+      flash[:error] = 'Failed to update payment transaction'
+    end
+
+    redirect_to admin_account_invoice_path(@invoice)
+  end
+
   protected
 
   def authorize_finance
     authorize! :read, Invoice
   end
 
+  def authorize_payment_gateway
+    return if payment_gateway_type.to_s =~ /stripe.*/
+    raise PaymentGatewayNotSupported
+  end
+
   def find_provider
     @provider = site_account
   end
+
+  def accessible_invoices
+    current_account.invoices.visible_for_buyer
+  end
+
+  def find_invoice
+    @invoice = accessible_invoices.find(params[:id])
```

Comment:
Same security issue without `Liquid::Drop`

Reply:
Right... and easy fix.